File: educative/js/find_permutation.js
// Given a string and a pattern, find out if the string contains any permutation of the pattern.
function _find_permutation(str, pattern) {
  const charFrequency = {};
  let start = 0;
  let matches = 0;
  let k = pattern.length;
  for (let i = 0; i < pattern.length; i++) {
    const char = pattern[i];
    if (!(char in charFrequency)) {
      charFrequency[char] = 0;
    }
    charFrequency[char]++;
  }

  for (let end = 0; end < str.length; end++) {
    const rightchar = str[end];
    if (rightchar in charFrequency) {
      charFrequency[rightchar] -= 1;
      if (charFrequency[rightchar] === 0) {
        matches++;
      }
    }

    if (matches === Object.keys(charFrequency).length) {
      return true;
    }

    if (end >= k - 1) {
      const leftchar = str[start];
      start++;
      if (leftchar in charFrequency) {
        if (charFrequency[leftchar] === 0) {
          matches = -1;
        }
        charFrequency[leftchar]++;
      }
    }
  }
  return false;
}

console.log(_find_permutation("bcdxabcdy", "bcdyabcdx"));
console.log(_find_permutation("odicf", "dc"));
console.log(_find_permutation("oidbcaf", "abc"));
console.log(_find_permutation("aaacb", "abc"));

File: educative/python/SlidingWindowPattern/find_permutation.py



def find_permutation(str, pattern):
  char_frequency = {}
  start = 0
  matches = 0
  for chr in pattern:
    if chr not in char_frequency:
      char_frequency[chr] = 0
    char_frequency[chr] += 1
  
  for end, _ in enumerate(str):
    rightchar = str[end]
    if rightchar in char_frequency:
      char_frequency[rightchar] -=1
      if char_frequency[rightchar] == 0:
        matches +=1
    if matches == len(char_frequency):
      return True
    
    if end >= len(pattern) -1:
      leftchar = str[start]
      start+=1
      if leftchar in char_frequency:
        if char_frequency[leftchar] ==0 :
          matches -=1
        char_frequency[leftchar] += 1
  return False
    

    
    




print(find_permutation("bcdxabcdy", "bcdyabcdx"))
print(find_permutation("odicf", "dc"));
print(find_permutation("oidbcaf", "abc"));
print(find_permutation("aaacb", "abc"));